src/PointwiseFunctions/Elasticity/ConstitutiveRelations/IsotropicHomogeneous.hpp

```diff
@@ -0,0 +1,104 @@
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <cstddef>
+#include <limits>
+
+#include "DataStructures/Tensor/TypeAliases.hpp"
+#include "Options/Options.hpp"
+#include "Utilities/TMPL.hpp"
+
+/// \cond
+namespace PUP {
+class er;
+}  // namespace PUP
+/// \endcond
+
+namespace Elasticity {
+/*!
+ * \brief Constitutive relations for elastic materials
+ */
+namespace ConstitutiveRelations {
+
+/*!
+ * \brief An isotropic and homogeneous material
+ *
+ * \details For an isotropic and homogeneous material the linear constitutive
+ * relation \f$T^{ij}=\chi^{ijkl}S_{kl}\f$ reduces to
+ * \f[
+ * \chi^{ijkl} = \lambda \delta^{ij}\delta^{kl} + \mu
+ * \left(\delta^{ik}\delta^{jl} + \delta^{il}\delta^{jk}\right) \\
+ * \implies \quad T^{ij} = \lambda \mathrm{Tr}(S) \delta^{ij} + 2\mu S^{ij}
+ * \f]
+ * with the _Lamé parameter_ \f$\lambda\f$ and the _rigidity_ (or _shear
+ * modulus_) \f$\mu\f$. In the parametrization chosen in this implementation we
+ * use the _incompressibility_ (or _bulk modulus_) \f[K=\lambda +
+ * \frac{2}{3}\mu\f] instead of the Lamé parameter. In this parametrization the
+ * stress-strain relation \f[ T^{ij} = K \mathrm{Tr}(S) \delta^{ij} +
+ * 2\mu\left(S^{ij} - \frac{1}{3}\mathrm{Tr}(S)\delta^{ij}\right)
+ * \f]
+ * decomposes into a scalar and a traceless part. Parameters also often used in
+ * this context are the _Young's modulus_
+ * \f[E=\frac{9K\mu}{3K+\mu}=\frac{\mu(3\lambda+2\mu)}{\lambda+\mu}\f] and the
+ * _Poisson ratio_
+ * \f[\nu=\frac{3K-2\mu}{6K+2\mu}=\frac{\lambda}{2(\lambda+\mu)}\f].
+ * Inversely, these relations read:
+ * \f{align}
+ * K &=\frac{E}{3-6\nu} \\
+ * \lambda &=\frac{E\nu}{(1+\nu)(1-2\nu)} \\
+ * \mu &=\frac{E}{2(1+\nu)}
+ * \f}
+ */
+template <size_t Dim>
+class IsotropicHomogeneous {
+ public:
+  struct Incompressibility {
+    using type = double;
+    static constexpr OptionString help = {"The bulk modulus of the material"};
+  };
+
+  struct Rigidity {
+    using type = double;
+    static constexpr OptionString help = {"The shear modulus of the material"};
+  };
+
+  using options = tmpl::list<Incompressibility, Rigidity>;
+
+  static constexpr OptionString help = {
+      "A constitutive relation that describes an isotropic, homogeneous "
+      "material."};
+
+  IsotropicHomogeneous() = default;
+  IsotropicHomogeneous(const IsotropicHomogeneous&) = delete;
+  IsotropicHomogeneous& operator=(const IsotropicHomogeneous&) = delete;
+  IsotropicHomogeneous(IsotropicHomogeneous&&) = default;
+  IsotropicHomogeneous& operator=(IsotropicHomogeneous&&) = default;
+  ~IsotropicHomogeneous() = default;
+
+  IsotropicHomogeneous(double incompressibility, double rigidity) noexcept;
+
+  /// The strain-stress relation that describes the material properties
+  template <typename DataType, typename Frame>
+  tnsr::II<DataType, Dim, Frame> stress(
+      const tnsr::ii<DataType, Dim, Frame>& strain,
+      const tnsr::I<DataType, Dim, Frame>& x) const noexcept;
```

Comment:
Isn't it true that all you need is the strain and the elastic moduli to get the stress tensor everywhere? Why compute it at a single point?

Reply:
We may want to implement inhomogeneous constitutive relations, which would need this